### DISCUSSION TITLE
How does pure association work?

### DISCUSSION THREAD

**Discussion by ianhbell** [ClapeyronThermo/Clapeyron.jl#260] [General]
I am mystified by the implementation in ``_X_exact1`` for single site association: https://github.com/ClapeyronThermo/Clapeyron.jl/blob/2a3d6eae261393bde743f57fd358d9ee5b955665/src/models/SAFT/association.jl#L328 . It clearly works, as numerical tests comparing against the explicit solutions in Huang&Radosz demonstrate, but I don't follow the logic. Is this taken from a paper, or re-derived? If the former, which paper, and if the latter, how?

**Comment by pw0908** [comment]
How do you mean? This is just transforming the mass-action equation into a quadratic equation and solving that.
$X_{H} = (1+n_e\rho\Delta_{He} X_e)^{-1}$
$X_{e} = (1+n_H\rho\Delta_{eH} X_H)^{-1}$
Substitute:
$X_{H} = (1+n_e\rho\Delta_{He} (1+n_H\rho\Delta_{eH} X_H)^{-1})^{-1}$
Multiply numerator and denominator by $(1+n_H\rho\Delta_{eH} X_H)$:
$X_{H} = (1+n_H\rho\Delta_{eH} X_H)(1+n_e\rho\Delta_{eH}+n_H\rho\Delta_{He} X_H)^{-1}$
Multiply the denominator across:
$X_{H}(1+n_e\rho\Delta_{He}+n_H\rho\Delta_{eH} X_H)=(1+n_H\rho\Delta_{eH} X_H)$
Re-write as quadratic:
$n_H\rho\Delta_{eH} X_H^2+(1+n_e\rho\Delta_{He}-n_H\rho\Delta_{eH})X_H-1=0$

Other association networks can also be solved analytically, this is just the most-common one. I believe I derived somewhere that you can only solve association networks up to 5 sites analytically but that's not included in Clapeyron.

**Reply by pw0908** [reply]
@ianhbell Here is the generalised solution. You end up with a quadratic

**Reply by ianhbell** [reply]
The paper from Kraska (https://pubs.acs.org/doi/10.1021/ie980206x) really helped me get my head around this problem.

**Reply by pw0908** [reply]
Are you talking about deriving the equations or solving the system of equations? For the former, I would recommend Laffite's 2013 paper (still one of the best SAFT papers I've read). For solving... there are lecture slides for the course that were taught by Michelsen somewhere online.

**Reply by ianhbell** [reply]
Indeed Lafitte has your magical equation: 
<img width="627" alt="image" src="https://github.com/user-attachments/assets/fd9ebdde-06af-489b-af6d-0b0dce378bd8">

The simplification approaches all assume that you have only A-B interactions, but no A-A and B-B self association interactions, right. Also, they assume that for mixtures, an A site on molecule i is interchangable with an A site on molecule j? Which is required to do the simplifications in Kraska

**Reply by pw0908** [reply]
Yes, the ij,ab = ij,ba simplification is something the old SAFT equations used to do. In SAFT-VR (gamma) Mie, they remove this assumption. To me, this makes sense as the water-H-alcohol-e interactions are going to be very different from the water-e-alcohol-H interactions.

**Reply by ianhbell** [reply]
I'm with you. As no expert in the domain what you wrote does make sense. In Kraska he keeps the non-simplified version where ∆eH ≠ ∆He and now I see the point why

**Reply by ianhbell** [reply]
I worked this out in sympy:
```python
# Here is the algebraic solutions to the laws of mass-action for the simplified cases
# covered in Huang & Radosz
import sympy as sy

kappa_B, kappa_A, X_A, X_B = sy.symbols('kappa_B, kappa_A, X_A, X_B')

# Definitions of the equations to be solved
# In Eq(), the first arg is the LHS, second is RHS
eq1 = sy.Eq(X_B, 1/(1+kappa_A*X_A))
eq2 = sy.Eq(X_A, 1/(1+kappa_B*X_B))
        
# The solutions
solns = sy.solve([eq1, eq2], [X_A, X_B])
for soln in solns:
    for x in soln:
        display(x)
```
yielding
<img width="432" alt="image" src="https://github.com/user-attachments/assets/6d78bc71-231f-45f7-b3f7-eb837ce84974">

of which only one pair should be valid

**Comment by ianhbell** [comment]
Ah that's cool. I'm no expert in this association stuff, so this is all new to me. Any good reference for the mass-action equations?

**Reply by pw0908** [reply]
The mass-action equation is generically written as:
$X_{i,a} = (1+\sum_j \rho_j\sum_b n_{j,b}\Delta_{ij,ab}X_{j,b})^{-1}\quad \forall i\in 1,\dots,n_c\quad a\in 1,\dots, n_{sites,i}$
Most (good) SAFT papers should include some details regarding how this equation is obtained / solved.

**Reply by ianhbell** [reply]
Ah, ok, now I see it, sorry I guess I was being dense.

**Reply by ianhbell** [reply]
This discussion arose because we were trying to make sense of the equation in Langenbach and Enders for pure fluids: 
<img width="526" alt="image" src="https://github.com/ClapeyronThermo/Clapeyron.jl/assets/11966765/2221b48a-8498-4191-8a26-381a0bd65d11">

**Reply by pw0908** [reply]
That's a really weird way of writing it...

**Reply by ianhbell** [reply]
Yeah, and also, incorrect, because you actually need the site fractions rather than the mole fractions where x_j is in that equation.

**Reply by ianhbell** [reply]
> The mass-action equation is generically written as: Xi,a=(1+∑jρj∑bnj,bΔij,abXj,b)−1∀i∈1,…,nca∈1,…,nsites,i Most (good) SAFT papers should include some details regarding how this equation is obtained / solved.

Can you recommend a good SAFT paper? Every one I have read seems to no consider multiple site types, with a single one of each type, rather than multiplicities of sites like you have written

**Comment by ianhbell** [comment]
Thanks @pw0908 for your explanation.
